cpp/oneapi/dal/backend/primitives/rng/rng_engine.hpp

```diff
@@ -0,0 +1,165 @@
+/*******************************************************************************
+* Copyright 2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#pragma once
+
+#include <daal/include/algorithms/engines/mt2203/mt2203.h>
+
+#include "oneapi/dal/backend/primitives/rng/utils.hpp"
+
+namespace oneapi::dal::backend::primitives {
+
+template <typename Type, typename Size = std::int64_t>
+class rng {
+public:
+    rng() = default;
+    ~rng() = default;
+
+    void uniform(Size count, Type* dst, void* state, Type a, Type b) {
+        uniform_by_cpu<Type>(count, dst, state, a, b);
+    }
+
+    void uniform_without_replacement(Size count,
+                                     Type* dst,
+                                     Type* buffer,
+                                     void* state,
+                                     Type a,
+                                     Type b) {
+        uniform_without_replacement_by_cpu<Type>(count, dst, buffer, state, a, b);
+    }
+
+    template <typename T = Type, typename = std::enable_if_t<std::is_integral_v<T>>>
+    cl::sycl::event shuffle(Size count, Type* dst, void* state) {
+        Type idx[2];
+
+        for (Size i = 0; i < count; ++i) {
+            uniform_by_cpu<Type>(2, idx, state, 0, count);
+            std::swap(dst[idx[0]], dst[idx[1]]);
+        }
+
+        return cl::sycl::event{};
+    }
+
+private:
+    sycl::queue queue_;
+    daal::internal::RNGs<Type, daal::sse2> daal_rng_;
+};
+
+class engine {
+public:
+    engine() : engine_(daal::algorithms::engines::mt2203::Batch<>::create()) {
+        impl_ = dynamic_cast<daal::algorithms::engines::internal::BatchBaseImpl*>(engine_.get());
```

Comment:
Maybe check that engine_.get() != nullptr?

Reply:
if it is null ptr dynamic_cast will return nullptr and it will be caught by if(!impl_) below